### PR TITLE
TestFramework: Refactoring and cleanup

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Follow up of #8601 

Commit history shows the changes in isolation

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs [hunk 1]
```diff
@@ -99,40 +96,23 @@ public static class DiagnosticVerifier
...
             var supportedDiagnostics = analyzer
                 .SelectMany(x => x.SupportedDiagnostics.Select(d => d.Id))
-                .Concat(new[] { AD0001 })
-                .Select(x => new KeyValuePair<string, ReportDiagnostic>(x, Severity(x)))
-                .ToArray();
-
-            var ids = supportedDiagnostics.Select(x => x.Key).ToHashSet();
-
+                .ToImmutableDictionary(x => x, Severity)
+                .Add(AD0001, ReportDiagnostic.Error);
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs:97]: I cannot find a better way to do this but it's a bit hard to the eye making an immutable dictionary and then re-creating it to add one entry 😅 .

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs [hunk 2]
```diff
@@ -99,40 +96,23 @@ public static class DiagnosticVerifier
...
             var compilationOptions = compilation.Options.WithSpecificDiagnosticOptions(supportedDiagnostics);
             var analyzerOptions = string.IsNullOrWhiteSpace(additionalFilePath) ? null : AnalysisScaffolding.CreateOptions(additionalFilePath);
             var diagnostics = compilation
                 .WithOptions(compilationOptions)
                 .WithAnalyzers(analyzer.ToImmutableArray(), analyzerOptions)
                 .GetAllDiagnosticsAsync(default)
                 .Result
-                .Where(x => (x.Severity == DiagnosticSeverity.Error && x.Id != LineContinuationVB12) || ids.Contains(x.Id))   // No compiler info about new syntax or unused usings
-                .ToImmutableArray();
+                .Where(x => (x.Severity == DiagnosticSeverity.Error && x.Id != LineContinuationVB12) || supportedDiagnostics.ContainsKey(x.Id));   // No compiler info about new syntax or unused usings
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs:105]: `LineContinuationVB12` what is this?

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM 
One question/comment.

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs:97]: This is the way immutable directories are build. They always get created then then entries are added as a AVL tree (yes, immutable dictionary is not a dictionary, but a tree)

Another example:
https://github.com/SonarSource/sonar-dotnet/blob/48f9399972fefba70fca547dcfead9a1e0ecb0ae/analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs#L41

**Pull request merged**